### DISCUSSION TITLE
Unable to fetch module versions from Gitlab Infrastructure Registry

### DISCUSSION THREAD

**Discussion by TuneableSloth** [renovatebot/renovate#18860] [Request Help] [answered]
**How are you running Renovate?**
Self-hosted

**If you're self-hosting Renovate, tell us what version of Renovate you run.**
renovate/renovate:34.20.0 (Helm Chart)

**If you're self-hosting Renovate, select which platform you are using.**
Kubernetes

**If you're self-hosting Renovate, tell us what version of the platform you run.**
1.23.12

**Relevant debug logs**

```
{
    "packageFile": "<REDACTED>",
    "deps": [
        {
            "currentValue": "1.0.0",
            "registryUrls": [
                "https://gitlab.com"
            ],
            "depType": "module",
            "depName": "gitlab.com/example/k8s-cluster/azure",
            "datasource": "terraform-module",
            "depIndex": 0,
            "updates": [],
            "warnings": [
                {
                    "topic": "gitlab.com/example/k8s-cluster/azure",
                    "message": "Failed to look up terraform-module dependency gitlab.com/example/k8s-cluster/azure"
                }
            ],
            "versioning": "hashicorp"
        }
    ]
}
```

However, the http statistics show the url is reachable.

`https://gitlab.com/api/v4/packages/terraform/modules/v1/example/k8s-cluster/azure/versions (GET,200)": 1,`

GET call to https://gitlab.com/api/v4/packages/terraform/modules/v1/example/k8s-cluster/azure/versions fetches all the available version.

It appears to be resolved as per this [issue](https://github.com/renovatebot/renovate/issues/14602).

**Have you created a minimal reproduction repository?**
No reproduction repository

Similar to https://gitlab.com/renovate-issue-14602

**Repository Config**
```
{
  "assigneesFromCodeOwners": true,
  "assigneesSampleSize": 2,
  "enabled": true,
  "hostRules": [
    {
      "hostType": "terraform-module",
      "matchHost": "https://gitlab.com/example/",
      "token": "${{ process.env.RENOVATE_TF_TOKEN }}"
    }
  ],
  "packageRules": [
    {
      "description": "Check for module from infrastructure registry",
      "matchPackageNames": [
        "gitlab.com/example/k8s-cluster/azure"
      ],
      "registryUrls": [
        "https://gitlab.com/example/"
      ],
      "addLabels": [
        "renovate-tf-module-update"
      ]
    }
  ],
  "rangeStrategy": "bump",
  "rebaseWhen": "behind-base-branch"
}
```

**Question**
Considering the platform that I am running on, is there any additional configuration required?

**Comment by viceice** [comment]
Can you check the logs for other requests for the terraform registry? or any failed http request?

**Reply by TuneableSloth** [reply]
Checked, no failed requests under `http statistics`. There are similar warnings for (_like one mentioned in logs above_) for other modules. The following call is a success as well: `"https://gitlab.com/.well-known/terraform.json (GET,200)": 1`.

**Comment by MindTooth** [comment]
I have the same problem, and been having it for some time.  What can I do to provide more information?

GitLab 15.5.  Renovate 34.23.1.

**Reply by MindTooth** [reply]
Yeah.  My own fault for not adding a host rule for how to access using a token it seems.  Setting the project to public, Renovate was able to access the module registry and start opening merge requests.

**Reply by TuneableSloth** [reply]
Would you be able to share the host rule if it is different from what I have configured? Also, if I may ask, what were the permissions set for the token?

**Reply by MindTooth** [reply]
I skipped the token step when I set the project public.  But I see now that you're using gitlab.com.

I'll come back to you when I've testet more by closing down the project again and start using tokens instead.

**Reply by TuneableSloth** [reply]
Unfortunately, I can't change my project's visibility to public. But thanks, looking forward to your inputs.

**Comment by TuneableSloth** [comment, marked as answer]
We happened to add the [host rules](https://github.com/renovatebot/renovate/discussions/18860#:~:text=2%2C%0A%20%20%22enabled%22%3A%20true%2C%0A%20%20%22-,hostRules,-%22%3A%20%5B%0A%20%20%20%20%7B%0A%20%20%20%20%20%20%22hostType%22%3A%20%22terraform%2Dmodule) to bot's config under secrets e.g - `(TERRAFORM__MODULE_GITLAB_COM_TOKEN)` and got it working. Reference [here](https://docs.renovatebot.com/self-hosted-configuration/#:~:text=to%20true.-,detectHostRulesFromEnv,-%C2%B6).

Thanks for taking out time and checking!

**Reply by MindTooth** [reply]
Do you have an example? 😝  I've been dealing with this for over a week, and I can't wrap my head around how Renovate works against a self-hosted GitLab.  Been toying with `hostRules`, and setting one, breaks the other (`terraform-module`, `git-tags`), and vice versa.  I'm looking for all help I can get atm.

**Reply by TuneableSloth** [reply]
No need to configure the `hostRules` for the _Infrastructure Module Registry_. Just specifying the `detectHostRulesFromEnv = true` in the bot's server config and specifying the environment variable `(TERRAFORM__MODULE_GITLAB_COM_TOKEN)` should work. The token value should have the usual access (suggest to start with a token with full access) on the repository.
